### PR TITLE
added wandb support

### PR DESCRIPTION
**Pull request by enochcluk**
**What does this PR do? Please describe:**
A summary of the change or the issue that is fixed.

Adds WandB support

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

**Instructions for use:**

conda install wandb from your conda environment
use wandb login from terminal or wandb.login() in python and input your API credentials (https://docs.wandb.ai/quickstart)
to set custom project name, add flag --config wandb_project=wandb_project_name to bash script

### CODE REVIEW

File: src/fairseq2/metrics/recorder.py [hunk 1]
```diff
@@ -474,5 +472,58 @@ def _get_writer(self, run: str) -> SummaryWriter | None:
...
+    def __init__(self, log_dir: Path) -> None:
+        """
+        :param log_dir:
+            The base directory under which to store the WandB files.
+        """
+        if not has_wandb:
+            log = get_log_writer(__name__)
+            log.warning(
+                "wandb not found. Please install it with `conda install wandb`."
+            )  # also need to wandb.login()
```
Comment (cbalioglu) [inline src/fairseq2/metrics/recorder.py:499]: What does "also need to wandb.login()" mean?

File: src/fairseq2/metrics/recorder.py [hunk 2]
```diff
@@ -474,5 +472,58 @@ def _get_writer(self, run: str) -> SummaryWriter | None:
...
+
+        self._log_dir = log_dir
+        path = Path(log_dir)
```
Comment (cbalioglu) [inline src/fairseq2/metrics/recorder.py:502]: `log_dir` is already a Path.

File: src/fairseq2/metrics/recorder.py [hunk 3]
```diff
@@ -474,5 +472,58 @@ def _get_writer(self, run: str) -> SummaryWriter | None:
...
+        path_components = path.parts
+        try:
+            self.project = path_components[-2]
```
Comment (cbalioglu) [inline src/fairseq2/metrics/recorder.py:505]: I think the project name should be passed as a parameter to the constructor.

**Comment by enochcluk** [comment]
this is not ready to merge yet, there is an issue with obj structure to config

**Pull request merged**